### ISSUE TITLE
Allow what-if to be performed without write permissions

### ISSUE THREAD

**Issue by michaeljfarr** [Azure/arm-template-whatif#135]
**Describe the bug**
Currently if I run "az deployment group what-if" on a subscription that I have read-only access to I receive an Authorization failed error.  Is this expected?

**To Reproduce**
1.Create a subscription that you do no have full permission to: $subscription
2.az deployment group what-if --resource-group $resourceGroup --template-file .\$resourceGroup\$resourceGroup.json --subscription $subscription --parameters .\$resourceGroup\Parameters_Development.json
3. The initial error looks something like this
ValidationError: InvalidTemplateDeployment - Deployment failed with multiple errors: 'Authorization failed for template resource 'bla-uat/web' of type 'Microsoft.Web/sites/config'. The client 'me@foo.com' with object id '66c5...' does not have permission to perform action 'Microsoft.Web/sites/config/write' at scope '/subscriptions/4d3.../resourceGroups/Digproviders/MitalStatus/providers/Microsoft.Web/sites/bla-uat/config/web'.

**Expected behavior**
No errors - assuming I have read access to all resources within the subscription

**Screenshots**
N/A

**Client** 
Powershell + azure-cli 2.12.1

**Additional context**
N/A

**Comment by alex-frankel** [comment]
Unfortunately, we don't have a way to run this with just Read permissions. You need `microsoft.resources/deployments/write` in order to run the /validate or /what-if API today.

**Comment by gallivantor** [comment]
Is this something that can be addressed? The what-if feature has great potential for all developers to use to see what effect their changes will have in Production, even if they don't have access to run it on Production themselves?

**Comment by alex-frankel** [comment]
> see what effect their changes will have in Production, even if they don't have access to run it on Production themselves

This scenario makes perfect sense, and I agree we should enable it. I have to imagine we can do this, but will need @majastrz  to comment.

**Comment by majastrz** [comment]
The minimal permissions you need to run What-If are:
- `Microsoft.Resources/deployments/whatIf/action` - allows making the POST /whatIf request 
- `Microsoft.Resources/subscriptions/operationresults/read` - allows reading the results
- Read permissions on the resources that are involved in the template and that are present in the RG.

Generally speaking, assigning all the permissions from the `Reader` role plus `Microsoft.Resources/deployments/whatIf/action` should work.

**Comment by michaeljfarr** [comment]
I've checked as well as I can that the user has the three permissions mentioned.  I'd be happy to confirm anything else if needed.

**Comment by gallivantor** [comment]
@majastrz  any ideas why your suggested permissions were not enough?

The error message says:
`The client 'me@foo.com' with object id '66c5...' does not have permission to perform action 'Microsoft.Web/sites/config/write' at scope XXX`
which implies it needs the write permission?

**Comment by majastrz** [comment]
The first step in What-If is to execute the template validation logic, which also performs something we call linked access checks (validation of permissions to resources whose IDs appear in properties of the another resource)

It looks like we have a bug in What-If because those permissions are being added to required what-if permisisons even though a normal deployment isn't being done by What-If. We will investigate how to fix this.

**Comment by SorenHartvigJensen** [comment]
> The first step in What-If is to execute the template validation logic, which also performs something we call linked access checks (validation of permissions to resources whose IDs appear in properties of the another resource)
> 
> It looks like we have a bug in What-If because those permissions are being added to required what-if permisisons even though a normal deployment isn't being done by What-If. We will investigate how to fix this.

Any updates?

We are using what-if in our pipelines, as a check before actually deploying to production, and it would be greatly appreciated if it was possible to do this without write permissions to production.

**Comment by adamglittle** [comment]
Just wanted to add another voice to this - being able to have a what-if pipeline that can run prior to other gates would be very helpful but we can't do this if the permissions are the same as an actual deploy. Is there any progress on the fix?

**Comment by matsest** [comment]
I guess this is not highly prioritized (since it's not been updated for more than over a year).. Do you have any information about if this bug is being investigated further @majastrz? Is a custom role still a recommended approach here?

FYI: this is something several customers I've worked with who's adopting Bicep over the last year has asked about @alex-frankel :)

**Comment by adamglittle** [comment]
Yep this is still a requirement for us. Disappointing to see the lack of progress

**Comment by alex-frankel** [comment]
There is no planned or active development on this specific issue due to a variety of resourcing limitations. We are starting to think about a what-if vNext might look like that more holistically addresses issues like this and some of the other frequently discussed bicep issues like noise and short-circuiting with bicep modules.

Unfortunately, that means nothing will move on this in the short term.

**Comment by Kaloszer** [comment]
+1, would be nice to have

**Comment by oramoss** [comment]
+1, really would like to see this.

**Comment by BasSmuldersCZ** [comment]
+1, really would like to see this.

**Comment by thomas-aure** [comment]
Hey! It would have been nice to see some plans from Microsoft regarding the current challenges with using ARM/Bicep as a method of deployment in a corporate setting. Specifically our concerns are with these two issues: 
1. What-If unpredictable short-circuiting issue.  
2. What-If permissions required. 

If the plan is to address some/any of these challenges in deploymentstacks or other solutions it would be nice to get some insights on this. :-)

**Comment by alex-frankel** [comment]
For 1: We are planning to fix this issue (tracked in #157), but this is the sort of challenge that we will only fully understand once the work has started. At the very least, we should be able to emit a warning/notice when short-circuiting happens. We will know more once we start work on it in the next month or two.

For 2: We have started looking into this as well, but this one is going to be challenging. Essentially what happens with what-if is, in addition to performing the GET request, we also run preflight, which is handled by the Resource Provider. The RP assumes you have write permissions when doing this check, which could have unpredictable escalation of privilege consequences. We need to come up with a safe way of reducing the permissions.

Both of these fixes will be independent of Deployment Stacks.

**Comment by matsest** [comment]
To follow up from the Bicep community meeting - any updates on this? 🤞 @alex-frankel

**Comment by haochuanwei** [comment]
+1, in many environments write permissions require a certain level of elevation that should be avoided whenever possible.
Not being able to run what-if in a CI environment is a major pain point.

**Comment by visschersm** [comment]
+1 I don't know if this already resolved but does not seem like it.
We are currently struggling with the permissions to only open up what is needed.
Would be nice to do this without permissions to write anything.

**Comment by shigeyf** [comment]
@alex-frankel

Do you have any updates on this?
I have the same problem on this issue, while Terraform plan does not require any "write" permissions.
https://github.com/Azure/bicep/discussions/17103

If my understanding is correct, on Az PowerShell module, there seems to be a resolution for this with ValidationLevel parameter (with ProviderNoRbac) (https://github.com/Azure/azure-powershell/pull/27352).
If so, are there any same (or similar) updates on Az CLI (python-based tool)?

Thanks in advance for your any thoughts or comments.

**Comment by aperezmicrosoft** [comment]
Hello @shigeyf and everyone - for this update you can now access this capability in PowerShell. Support for CLI is currently in rollout process and will be available in version 2.74.0 of CLI. We will update this thread next week with an update on that release. Thanks!

**Comment by matsest** [comment]
The release notes from the PowerShell team was not very verbose about this addition, didn't catch it at all - but it actually works! Although there was a gotcha with permissions needed (currently a custom role is needed for least-privileged approach).

I made a write-up with a demo of the new feature here for anyone interested https://github.com/matsest/bicep-deploy-whatif-without-write-demo

Not every day you find a solution to a 5 year old issue - great work Bicep/ARM team! 💪

**Comment by SorenHartvigJensen** [comment]
> Hello [@shigeyf](https://github.com/shigeyf) and everyone - for this update you can now access this capability in PowerShell. Support for CLI is currently in rollout process and will be available in version 2.74.0 of CLI. We will update this thread next week with an update on that release. Thanks!

I see that Azure CLI 2.74.0 has now been released but there is nothing in the release notes regarding support for validation level or any other changes to what-if?

https://github.com/MicrosoftDocs/azure-docs-cli/blob/main/docs-ref-conceptual/Latest-version/release-notes-azure-cli.md

**Comment by SorenHartvigJensen** [comment]
Did anyone find any documentation on the implementation of validationLevel on Azure CLI?

**Comment by matsest** [comment]
Does not seem like it's implemented yet, even in the 2.75 version released today 🤷 

(Available in the Azure PowerShell module as well as the https://github.com/Azure/bicep-deploy action)

Edit: seems like it is process with https://github.com/Azure/azure-cli/pull/31747

**Comment by jeskew** [comment]
@SorenHartvigJensen apologies for the delay on support in the CLI. I thought there was an automatic binding of API parameters to CLI commands, and @azcloudfarmer's message above was based on my incorrect assumption. The CLI PR linked to in the previous comment will take care of this

**Comment by SorenHartvigJensen** [comment]
> [@SorenHartvigJensen](https://github.com/SorenHartvigJensen) apologies for the delay on support in the CLI. I thought there was an automatic binding of API parameters to CLI commands, and [@azcloudfarmer](https://github.com/azcloudfarmer)'s message above was based on my incorrect assumption. The CLI PR linked to in the previous comment will take care of this

Amazing. So next release should be the one?

**Comment by torreymicrosoft** [comment]
Thanks for sticking with this one, and sorry for how long it ran. Picking it back up as one of the older open issues for the ARM Deployments team to review.

The ask here was to run what-if against a scope where you only have read access, instead of getting an authorization failure on something like `Microsoft.Web/sites/config/write`. That is now shipped on both clients, and the CLI half is the piece this thread was still waiting on.

I checked the Azure CLI I have here, 2.87.0, and `az deployment group what-if --help` lists:

```
--validation-level : The deployment validation level. May be set to "Provider" (the default),
                     "Template", or "ProviderNoRbac". ... With a validation level of
                     "providerNoRbac", ARM will perform full validation but only check for
                     read permissions on each resource.
                     Allowed values: Provider, ProviderNoRbac, Template.
```

`az deployment sub what-if --help` carries the same parameter, so it is not resource group only.

So the usage is:

```
az deployment group what-if \
  --resource-group my-rg \
  --template-file main.bicep \
  --parameters @params.json \
  --validation-level ProviderNoRbac
```

@SorenHartvigJensen, that answers your last question directly: it did land, and it is in the CLI now rather than pending. @matsest, thank you for the write up and the demo repo, that was the clearest confirmation on this thread that the PowerShell side worked, and your note about needing a custom role for a genuinely least privileged setup is the right caveat for anyone arriving here.

I am going to close this out as fixed. If `ProviderNoRbac` still trips an authorization failure for you on a scope where you hold read access, reopen with the error and a correlation ID and I will take it from there.
